File: Aed's_II/Verde/Java/Game.java
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#define MAX_STR 200
#define MAX_LISTA 1000
#define MAX_VETOR 100

static char *meu_strdup(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s) + 1;
    char *p = (char*)malloc(n);
    if (p) memcpy(p, s, n);
    return p;
}

static int meu_strcasecmp(const char *a, const char *b) {
    if (!a || !b) return (a==b)?0:(a?1:-1);
    while (*a && *b) {
        unsigned char ca = (unsigned char) tolower((unsigned char)*a);
        unsigned char cb = (unsigned char) tolower((unsigned char)*b);
        if (ca != cb) return ca - cb;
        a++; b++;
    }
    return (unsigned char) tolower((unsigned char)*a) - (unsigned char) tolower((unsigned char)*b);
}

typedef struct Jogo {
    int identificadorApp;
    char nome[MAX_STR];
    char dataLancamento[MAX_STR];
    int proprietariosEstimados;
    double preco;
    char idiomasSuportados[MAX_VETOR][MAX_STR];
    int numIdiomas;
    int pontuacaoMetacritic;
    double pontuacaoUsuario;
    int conquistas;
    char editoras[MAX_VETOR][MAX_STR];
    int numEditoras;
    char desenvolvedores[MAX_VETOR][MAX_STR];
    int numDesenvolvedores;
    char categorias[MAX_VETOR][MAX_STR];
    int numCategorias;
    char generos[MAX_VETOR][MAX_STR];
    int numGeneros;
    char etiquetas[MAX_VETOR][MAX_STR];
    int numEtiquetas;

    char proprietariosEstimadosBruto[MAX_STR];
    char precoBruto[MAX_STR];
    char pontuacaoUsuarioBruta[MAX_STR];
    char conquistasBruto[MAX_STR];
} Jogo;

long comparacoes = 0;
long movimentacoes = 0;

int parseInteiroDigitos(const char *s) {
    char tmp[MAX_STR]; int j=0;
    if (!s) return 0;
    for (int i=0; s[i]; i++) {
        if (isdigit((unsigned char)s[i])) tmp[j++]=s[i];
    }
    tmp[j]='\0';
    return j>0?atoi(tmp):0;
}

double parseDoubleSeguro(const char *s) {
    if (!s) return 0.0;
    if (meu_strcasecmp(s, "Free to Play")==0) return 0.0;
    return atof(s);
}

int dividirParaVetor(char destino[MAX_VETOR][MAX_STR], const char *s) {
    if (!s || strlen(s)==0) return 0;
    char copia[MAX_STR*2]; strncpy(copia, s, sizeof(copia)-1); copia[sizeof(copia)-1]='\0';
    for (int i=0; copia[i]; i++) if (copia[i]=='['||copia[i]==']'||copia[i]=='\"' || copia[i]=='\'' ) copia[i]=' ';
    int contador=0;
    char *tok = strtok(copia, ",");
    while (tok && contador<MAX_VETOR) {
        while (*tok==' ') tok++;
        strncpy(destino[contador], tok, MAX_STR-1); destino[contador][MAX_STR-1]='\0';
        contador++;
        tok = strtok(NULL, ",");
    }
    return contador;
}

void formatarDados(Jogo *jogo) {
    if (strlen(jogo->dataLancamento)==0) strcpy(jogo->dataLancamento, "01/01/1970");
    jogo->proprietariosEstimados = parseInteiroDigitos(jogo->proprietariosEstimadosBruto);
    jogo->preco = parseDoubleSeguro(jogo->precoBruto);
    if (jogo->pontuacaoUsuarioBruta[0]==0 || meu_strcasecmp(jogo->pontuacaoUsuarioBruta, "tbd")==0) jogo->pontuacaoUsuario=-1.0;
    else jogo->pontuacaoUsuario = atof(jogo->pontuacaoUsuarioBruta);
    if (jogo->pontuacaoMetacritic==0) jogo->pontuacaoMetacritic=-1;
    jogo->conquistas = parseInteiroDigitos(jogo->conquistasBruto);
}

void imprimirJogo(const Jogo *jogo) {
    printf("=> %d ## %s ## %s ## %d ## %.2f ## ", jogo->identificadorApp, jogo->nome, jogo->dataLancamento, jogo->proprietariosEstimados, jogo->preco);
    printf("[");
    for (int i=0;i<jogo->numIdiomas;i++) { printf("%s", jogo->idiomasSuportados[i]); if (i<jogo->numIdiomas-1) printf(", "); }
    printf("] ## %d ## %.1f ## %d ## ", jogo->pontuacaoMetacritic, jogo->pontuacaoUsuario, jogo->conquistas);
    printf("["); for (int i=0;i<jogo->numDesenvolvedores;i++){ printf("%s", jogo->desenvolvedores[i]); if (i<jogo->numDesenvolvedores-1) printf(", "); } printf("] ## ");
    printf("["); for (int i=0;i<jogo->numEditoras;i++){ printf("%s", jogo->editoras[i]); if (i<jogo->numEditoras-1) printf(", "); } printf("] ## ");
    printf("["); for (int i=0;i<jogo->numCategorias;i++){ printf("%s", jogo->categorias[i]); if (i<jogo->numCategorias-1) printf(", "); } printf("] ## ");
    printf("["); for (int i=0;i<jogo->numGeneros;i++){ printf("%s", jogo->generos[i]); if (i<jogo->numGeneros-1) printf(", "); } printf("] ## ");
    printf("["); for (int i=0;i<jogo->numEtiquetas;i++){ printf("%s", jogo->etiquetas[i]); if (i<jogo->numEtiquetas-1) printf(", "); } printf("] ##\n");
}

int compararJogo(const Jogo *a, const Jogo *b) {
    int diaA=1,mesA=1,anoA=1970; int diaB=1,mesB=1,anoB=1970;
    if (a->dataLancamento[0]) sscanf(a->dataLancamento, "%d/%d/%d", &diaA, &mesA, &anoA);
    if (b->dataLancamento[0]) sscanf(b->dataLancamento, "%d/%d/%d", &diaB, &mesB, &anoB);
    comparacoes++; if (anoA<anoB) return -1; comparacoes++; if (anoA>anoB) return 1;
    comparacoes++; if (mesA<mesB) return -1; comparacoes++; if (mesA>mesB) return 1;
    comparacoes++; if (diaA<diaB) return -1; comparacoes++; if (diaA>diaB) return 1;
    comparacoes++; if (a->identificadorApp < b->identificadorApp) return -1; comparacoes++; if (a->identificadorApp > b->identificadorApp) return 1;
    return 0;
}

int particionar(Jogo arr[], int inicio, int fim) {
    Jogo pivo = arr[fim];
    int i = inicio - 1;
    for (int j=inicio;j<fim;j++) {
        if (compararJogo(&arr[j], &pivo) < 0) {
            i++;
            Jogo temp = arr[i]; arr[i]=arr[j]; arr[j]=temp; movimentacoes++;
        }
    }
    Jogo temp = arr[i+1]; arr[i+1]=arr[fim]; arr[fim]=temp; movimentacoes++;
    return i+1;
}

void quicksort(Jogo arr[], int inicio, int fim) {
    if (inicio<fim) {
        int pi = particionar(arr, inicio, fim);
        quicksort(arr, inicio, pi-1);
        quicksort(arr, pi+1, fim);
    }
}

int lerCSV(const char *caminho, Jogo lista[], int max) {
    FILE *fp = fopen(caminho, "r");
    if (!fp) return 0;
    char linha[4096];
    if (!fgets(linha, sizeof(linha), fp)) { fclose(fp); return 0; }
    int contador=0;
    while (fgets(linha, sizeof(linha), fp) && contador<max) {
        linha[strcspn(linha, "\n")]='\0';
        if (strlen(linha)==0) continue;
        
        char *campos[14];
        int idxCampos=0;
        char buf[4096]; int idxBuf=0; int entreAspas=0;
        for (int i=0; linha[i]; i++) {
            char c=linha[i];
            if (c=='"') { entreAspas = !entreAspas; }
            if (c==',' && !entreAspas) { buf[idxBuf]='\0'; campos[idxCampos++]=meu_strdup(buf); idxBuf=0; }
            else { buf[idxBuf++]=c; }
        }
        buf[idxBuf]='\0'; if (idxCampos<14) campos[idxCampos++]=meu_strdup(buf);
        
        if (idxCampos>=14) {
            Jogo jogo; memset(&jogo,0,sizeof(Jogo));
            jogo.identificadorApp = atoi(campos[0]);
            strncpy(jogo.nome, campos[1], MAX_STR-1);
            strncpy(jogo.dataLancamento, campos[2], MAX_STR-1);
            strncpy(jogo.proprietariosEstimadosBruto, campos[3], MAX_STR-1);
            strncpy(jogo.precoBruto, campos[4], MAX_STR-1);
            jogo.numIdiomas = dividirParaVetor(jogo.idiomasSuportados, campos[5]);
            jogo.pontuacaoMetacritic = atoi(campos[6]);
            strncpy(jogo.pontuacaoUsuarioBruta, campos[7], MAX_STR-1);
            strncpy(jogo.conquistasBruto, campos[8], MAX_STR-1);
            jogo.numEditoras = dividirParaVetor(jogo.editoras, campos[9]);
            jogo.numDesenvolvedores = dividirParaVetor(jogo.desenvolvedores, campos[10]);
            jogo.numCategorias = dividirParaVetor(jogo.categorias, campos[11]);
            jogo.numGeneros = dividirParaVetor(jogo.generos, campos[12]);
            jogo.numEtiquetas = dividirParaVetor(jogo.etiquetas, campos[13]);
            formatarDados(&jogo);
            lista[contador++]=jogo;
        }
        for (int k=0;k<idxCampos;k++) free(campos[k]);
    }
    fclose(fp);
    return contador;
}

int buscarPorId(Jogo lista[], int numItens, int id) {
    for (int i=0;i<numItens;i++) if (lista[i].identificadorApp==id) return i;
    return -1;
}

int main() {
    char *caminhos[] = {"/tmp/games.csv" };
    Jogo bancoDeDados[MAX_LISTA]; int total=0;
    for (int p=0;p<4 && total==0;p++) total = lerCSV(caminhos[p], bancoDeDados, MAX_LISTA);
    if (total==0) { fprintf(stderr, "games.csv not found\n"); return 1; }

    Jogo selecionados[MAX_LISTA]; int numSelecionados=0;
    char linha[128];
    while (fgets(linha, sizeof(linha), stdin)) {
        linha[strcspn(linha, "\n")]='\0'; if (strcmp(linha,"FIM")==0) break; if (strlen(linha)==0) continue;
        int id = atoi(linha);
        int indice = buscarPorId(bancoDeDados, total, id);
        if (indice>=0) selecionados[numSelecionados++]=bancoDeDados[indice];
    }

    comparacoes=0; movimentacoes=0;
    clock_t inicioTempo = clock();
    if (numSelecionados>0) quicksort(selecionados, 0, numSelecionados-1);
    clock_t fimTempo = clock();
    long tempoExecucao = (long)((fimTempo-inicioTempo)*1000000/ CLOCKS_PER_SEC);

    for (int i=0;i<numSelecionados;i++) imprimirJogo(&selecionados[i]);

    FILE *arquivoLog = fopen("845963_quicksort.txt","w");
    if (arquivoLog) { fprintf(arquivoLog, "845963\t%ld\t%ld\t%ld\n", comparacoes, movimentacoes, tempoExecucao); fclose(arquivoLog); }

    return 0;
}

File: Aed's_II/Verde/C/Game.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#define MAX_LINE 4096
#define MAX_GAMES 5000
#define MAX_STRING 1024
#define MATRICULA "845963"

#ifdef _WIN32
#include <strings.h>
#ifndef strcasecmp
#define strcasecmp _stricmp
#endif
#ifndef strncasecmp
#define strncasecmp _strnicmp
#endif
#endif

typedef struct {
    int id;
    char name[MAX_STRING];
    char releaseDate[12];
    int estimatedOwners;
    double price;
    char supportedLanguages[MAX_STRING];
    int metacriticScore;
    double userScore;
    int achievements;
    char publishers[MAX_STRING];
    char developers[MAX_STRING];
    char categories[MAX_STRING];
    char genres[MAX_STRING];
    char tags[MAX_STRING];
} Game;

// Remove espaços do início e fim da string
void trim(char *str) {
    if (!str || *str == '\0') return;
    char *start = str;
    char *end = str + strlen(str) - 1;
    while (isspace((unsigned char)*start)) start++;
    while (end > start && isspace((unsigned char)*end)) end--;
    size_t len = (end - start) + 1;
    memmove(str, start, len);
    str[len] = '\0';
}

// Remove caracteres indesejados
void removeChars(char *str, const char *charsToRemove) {
    char *src = str, *dst = str;
    while (*src) {
        if (strchr(charsToRemove, *src) == NULL) {
            *dst++ = *src;
        }
        src++;
    }
    *dst = '\0';
}

// Retorna número do mês
int getMonthNumber(char *month) {
    char *months[] = { "Jan","Feb","Mar","Apr","May","Jun",
                       "Jul","Aug","Sep","Oct","Nov","Dec" };
    char *fullMonths[] = { "January","February","March","April","May","June",
                           "July","August","September","October","November","December" };
    for (int i = 0; i < 12; i++) {
        if (strncasecmp(month, months[i], 3) == 0 || strcasecmp(month, fullMonths[i]) == 0)
            return i + 1;
    }
    return 0;
}

// Converte datas para dd/mm/aaaa
void parseDate(char *input, char *output) {
    char inputCopy[MAX_STRING];
    strcpy(inputCopy, input);
    trim(inputCopy);

    int day = 0, month = 0, year = 0;
    char monthStr[50];

    if (sscanf(inputCopy, "%s %d, %d", monthStr, &day, &year) == 3) {
        month = getMonthNumber(monthStr);
        if (month > 0) { sprintf(output, "%02d/%02d/%04d", day, month, year); return; }
    }
    if (sscanf(inputCopy, "%d-%d-%d", &year, &month, &day) == 3 && year > 1000) {
        sprintf(output, "%02d/%02d/%04d", day, month, year); return;
    }
    if (sscanf(inputCopy, "%d/%d/%d", &day, &month, &year) == 3 && day <= 31 && month <= 12) {
        sprintf(output, "%02d/%02d/%04d", day, month, year); return;
    }
    strcpy(output, "01/01/1970");
}

// Extrai número de donos estimados
int extractEstimatedOwners(char *raw) {
    char buffer[MAX_STRING]; int j = 0;
    for (int i = 0; raw[i] != '\0' && i < MAX_STRING - 1; i++)
        if (isdigit(raw[i])) buffer[j++] = raw[i];
    buffer[j] = '\0';
    return (j > 0) ? atoi(buffer) : 0;
}

// Extrai preço
double extractPrice(char *raw) {
    if (strcasecmp(raw, "Free to Play") == 0) return 0.0;
    return atof(raw);
}

// Extrai user score
double extractUserScore(char *raw) {
    if (strcasecmp(raw, "tbd") == 0) return -1.0;
    return atof(raw);
}

// Extrai achievements
int extractAchievements(char *raw) {
    trim(raw);
    return (strlen(raw) == 0) ? 0 : atoi(raw);
}

// Divide linha em campos
int parseLine(char *line, char fields[][MAX_STRING]) {
    int inQuotes = 0, fieldIndex = 0, bufferIndex = 0;
    for (int i = 0; line[i] != '\0' && fieldIndex < 20; i++) {
        char c = line[i];
        if (c == '"') inQuotes = !inQuotes;
        else if (c == ',' && !inQuotes) {
            fields[fieldIndex][bufferIndex] = '\0';
            trim(fields[fieldIndex]);
            fieldIndex++; bufferIndex = 0;
        } else {
            if (bufferIndex < MAX_STRING - 1) fields[fieldIndex][bufferIndex++] = c;
        }
    }
    fields[fieldIndex][bufferIndex] = '\0';
    trim(fields[fieldIndex]);
    return fieldIndex + 1;
}

// Processa campos em struct Game
void processGame(Game *game, char fields[][MAX_STRING]) {
    game->id = atoi(fields[0]);
    strncpy(game->name, fields[1], MAX_STRING - 1); game->name[MAX_STRING - 1] = '\0';
    parseDate(fields[2], game->releaseDate);
    game->estimatedOwners = extractEstimatedOwners(fields[3]);
    game->price = extractPrice(fields[4]);

    strncpy(game->supportedLanguages, fields[5], MAX_STRING - 1); 
    game->supportedLanguages[MAX_STRING - 1] = '\0';
    removeChars(game->supportedLanguages, "[]\"''"); trim(game->supportedLanguages);

    game->metacriticScore = atoi(fields[6]);
    if (game->metacriticScore == 0) game->metacriticScore = -1;

    game->userScore = extractUserScore(fields[7]);
    game->achievements = extractAchievements(fields[8]);

    strncpy(game->publishers, fields[9], MAX_STRING - 1); game->publishers[MAX_STRING - 1] = '\0';
    removeChars(game->publishers, "[]\"''"); trim(game->publishers);

    strncpy(game->developers, fields[10], MAX_STRING - 1); game->developers[MAX_STRING - 1] = '\0';
    removeChars(game->developers, "[]\"''"); trim(game->developers);

    strncpy(game->categories, fields[11], MAX_STRING - 1); game->categories[MAX_STRING - 1] = '\0';
    removeChars(game->categories, "[]\"''"); trim(game->categories);

    strncpy(game->genres, fields[12], MAX_STRING - 1); game->genres[MAX_STRING - 1] = '\0';
    removeChars(game->genres, "[]\"''"); trim(game->genres);

    strncpy(game->tags, fields[13], MAX_STRING - 1); game->tags[MAX_STRING - 1] = '\0';
    removeChars(game->tags, "[]\"''"); trim(game->tags);
}

// Formata e imprime Game
void printGame(Game *game) {
    char supportedLanguages[MAX_STRING], publishers[MAX_STRING], developers[MAX_STRING];
    char categories[MAX_STRING], genres[MAX_STRING], tags[MAX_STRING];

    #define formatList(dest, src) \
        { int j=0; for(int i=0; src[i]!='\0'&&j<MAX_STRING-2; i++){ dest[j++] = src[i]; if(src[i]==',') dest[j++]=' '; } dest[j]='\0'; }

    formatList(supportedLanguages, game->supportedLanguages);
    formatList(publishers, game->publishers);
    formatList(developers, game->developers);
    formatList(categories, game->categories);
    formatList(genres, game->genres);
    formatList(tags, game->tags);

    printf("=> %d ## %s ## %s ## %d ## %.2f ## [%s] ## %d ## %.1f ## %d ## [%s] ## [%s] ## [%s] ## [%s] ## [%s] ##\n",
        game->id, game->name, game->releaseDate, game->estimatedOwners, game->price,
        supportedLanguages, game->metacriticScore, game->userScore, game->achievements,
        publishers, developers, categories, genres, tags
    );
}

// Busca Game por ID
int findGameById(Game *games, int count, int targetId) {
    for(int i=0;i<count;i++) if(games[i].id==targetId) return i;
    return -1;
}

// Selection Sort por nome (case-insensitive)
void selectionSortByName(Game *games, int n, long long *comparisons, long long *movements) {
    for(int i=0;i<n-1;i++){
        int minIdx = i;
        for(int j=i+1;j<n;j++){
            (*comparisons)++;
            if(strcasecmp(games[j].name, games[minIdx].name)<0) minIdx=j;
        }
        if(minIdx!=i){
            Game tmp = games[i]; games[i]=games[minIdx]; games[minIdx]=tmp;
            (*movements)+=3;
        }
    }
}

int main() {
    const char *path = "/tmp/games.csv";
    Game *games = (Game *) malloc(MAX_GAMES*sizeof(Game));
    if(!games){ fprintf(stderr,"Erro ao alocar memoria\n"); return 1; }

    int gameCount=0;
    FILE *file = fopen(path,"r");
    if(!file){ fprintf(stderr,"Erro ao abrir arquivo: %s\n",path); free(games); return 1; }

    char line[MAX_LINE];
    if(fgets(line,MAX_LINE,file)==NULL){ fclose(file); free(games); return 1; }

    while(fgets(line,MAX_LINE,file) && gameCount<MAX_GAMES){
        line[strcspn(line,"\n")]='\0';
        if(strlen(line)==0) continue;
        char fields[20][MAX_STRING];
        int fieldCount=parseLine(line,fields);
        if(fieldCount>=14){ processGame(&games[gameCount],fields); gameCount++; }
    }
    fclose(file);

    Game *selected = (Game *) malloc(MAX_GAMES*sizeof(Game));
    if(!selected){ fprintf(stderr,"Erro ao alocar memória para selecionados\n"); free(games); return 1; }

    int selCount=0;
    char input[MAX_STRING];
    while(fgets(input,MAX_STRING,stdin)){
        input[strcspn(input,"\n")]='\0'; trim(input);
        if(strcasecmp(input,"FIM")==0) break;
        int id=atoi(input); int idx=findGameById(games,gameCount,id);
        if(idx!=-1) selected[selCount++]=games[idx];
    }

    long long comparisons=0, movements=0;
    clock_t start=clock();
    selectionSortByName(selected,selCount,&comparisons,&movements);
    clock_t end=clock();
    double elapsed=((double)(end-start))/CLOCKS_PER_SEC;

    char logName[256]; snprintf(logName,sizeof(logName),"%s_selecao.txt",MATRICULA);
    FILE *logFile=fopen(logName,"w");
    if(logFile){ fprintf(logFile,"%s\t%lld\t%lld\t%.6f\n",MATRICULA,comparisons,movements,elapsed); fclose(logFile); }

    for(int i=0;i<selCount;i++) printGame(&selected[i]);

    free(selected); free(games);
    return 0;
}
